File: aoc_2022/src/day1.ts
import { BunFile } from "bun";
import path from "path";

// read file
const path: string = "day1.input";

const file: BunFile = Bun.file(path);

const contents: string = await file.text();

// console.log(contents);

// split content by blankline

const blocks = contents.split("\n\n");

// map block index to total

const calories = blocks.map((block) => {
  return block.split("\n");
});

const totals: number[] = calories.map((block) => {
  return block.map((calorie) => {
    return +calorie; // or parseInt(calorie)
  })
    .filter((num) => !isNaN(num))
    .reduce((acc, current) => acc + current, 0);
});

// console.log(totals);
// return the max total
// const max = Math.max(...totals);
const sortedTotals = totals.sort((a, b) => b - a);
const max = sortedTotals[0];

console.log(max);

const sum3 = sortedTotals.slice(0, 3).reduce((a, b) => a + b, 0);
console.log(`part two: ${sum3}`);

File: aoc_2022/src/day1.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	s "strings"
)

var p = fmt.Println

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type elf struct {
	position int
	total    int
}

func sum(nums ...int) int {
	sum := 0
	for _, num := range nums {
		sum += num
	}
	return sum
}
func main() {
	// read entire file to mem
	data, err := os.ReadFile("day1.input")
	check(err)
	content := string(data)

	// split by blankline
	food := s.Split(content, "\n\n")

	calories := make([][]int, len(food))
	for i := range food {
		foods := food[i]
		newCals := s.Split(foods, "\n")
		mappedCals := make([]int, len(newCals))
		for j, val := range newCals {
			cal, err := strconv.Atoi(val)
			if err != nil {
				continue
			}
			mappedCals[j] = cal
		}
		calories[i] = append(calories[i], mappedCals...)
	}
	elves := make([]elf, len(calories))
	for i, elfCals := range calories {
		elves[i] = elf{
			position: i + 1,
			total:    sum(elfCals...),
		}
	}

	sort.Slice(elves, func(i, j int) bool {
		return elves[i].total > elves[j].total
	})

	fmt.Printf("the %dth elf that has the most calories: %d\n", elves[0].position, elves[0].total)

	var top3 []int
	for _, elve := range elves[0:3] {
		top3 = append(top3, elve.total)
	}
	sum3 := sum(top3...)
	fmt.Printf("the sum of top 3 is %d\n", sum3)
}
